src/deepLink/deepLinkService.ts

```diff
@@ -0,0 +1,403 @@
+		this._state = DeepLinkServiceStates.Idle;
+		this._disposables = [
+			container.uri.onUri((event: UriEvent) => {
+				if (event.type === UriTypes.DeepLink && this._state === DeepLinkServiceStates.Idle) {
+					if (!event.repoId || !event.linkType || !event.uri || !event.remoteUrl) {
+						void window.showErrorMessage(`Error resolving deep link: missing required properties.`);
+						return;
+					}
+
+					if (!Object.values(DeepLinkTypes).includes(event.linkType)) {
+						void window.showErrorMessage(`Error resolving deep link: unknown link type.`);
+						return;
+					}
+
+					if (event.linkType !== DeepLinkTypes.Remote && !event.targetId) {
+						void window.showErrorMessage(
+							`Error resolving deep link of type ${event.linkType}: no target id provided.`,
+						);
+						return;
+					}
+
+					this._repoId = event.repoId;
+					this._targetType = event.linkType;
+					this._uri = event.uri;
+					this._remoteUrl = event.remoteUrl;
+					this._targetId = event.targetId;
+
+					this._stateChange.fire({
+						state: this._state,
+						action: DeepLinkServiceActions.DeepLinkEventFired,
+					});
+				}
+			}),
+			this._stateChange.event(async (serviceStateChange: DeepLinkServiceStateChange) => {
+				await this.handleDeepLinkStateChange(serviceStateChange);
+			}),
+		];
+	}
+
+	dispose() {
+		this._disposables.forEach((disposable: Disposable) => void disposable.dispose());
+	}
+
+	async getShaForTarget(): Promise<string | undefined> {
+		if (!this._repo || !this._remote || this._targetType === DeepLinkTypes.Remote || !this._targetId) {
+			return undefined;
+		}
+
+		if (this._targetType === DeepLinkTypes.Branch) {
+			// Form the target branch name using the remote name and branch name
+			const branchName = `${this._remote.name}/${this._targetId}`;
+			const branch = await this._repo.getBranch(branchName);
+			if (branch) {
+				return branch.sha;
+			}
+
+			return undefined;
+		}
+
+		if (this._targetType === DeepLinkTypes.Tag) {
+			const tag = await this._repo.getTag(this._targetId);
+			if (tag) {
+				return tag.sha;
+			}
+
+			return undefined;
+		}
+
+		if (this._targetType === DeepLinkTypes.Commit) {
+			if (await this.container.git.validateReference(this._repo.path, this._targetId)) {
+				return this._targetId;
+			}
+
+			return undefined;
+		}
+
+		return undefined;
+	}
+
+	async handleDeepLinkStateChange(serviceStateChange: DeepLinkServiceStateChange) {
+		const { state, action, data } = serviceStateChange;
+		let nextState = this._transitionTable[state][action];
+		let nextData: any;
+		let nextAction: DeepLinkServiceAction = DeepLinkServiceActions.DeepLinkErrored;
+		if (!nextState) {
+			nextState = DeepLinkServiceStates.Idle;
+		}
+
+		this._state = nextState;
+		switch (nextState) {
+			case DeepLinkServiceStates.Idle:
+				if (action === DeepLinkServiceActions.DeepLinkResolved) {
+					void window.showInformationMessage(`Deep link resolved: ${this._uri?.toString()}`);
```

Comment:
I think the messages are helpful. Ultimately you would only see one for a deep link, but it would help signal to the user that the deep link processing is done (it has finished getting to its target) or give some context on why it failed so they aren't stuck figuring out why it failed. Also, in the case of bugs with the feature, reporters can distinguish whether they received an error on it or no messaging at all, which would indicate that it got stuck somewhere in processing.

We can definitely include discussion of the messaging in this PR review. What do you think of the current messaging?

Noted re: `else if`, will update it, thanks!

Reply:
Updated the `if`s in commit: [Short-circuit action checking in idle state case](https://github.com/gitkraken/vscode-gitlens/pull/2471/commits/6f67236c4ca6dca05ee708177e6b892c3aba3b22)